### PR TITLE
Prediction

### PR DESCRIPTION
**Pull request by samedii**
Template:
- Collate in model
- Preprocess in model
- Prediction output from model

### CODE REVIEW

File: template/{{cookiecutter.repository_name}}/{{cookiecutter.package_name}}/architecture/prediction.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        for index in range(len(self)):
+            yield self[index]
+
+    def loss(self, class_names):
+        targets = torch.tensor(
+            [class_index(class_name) for class_name in class_names],
+        ).to(self.logits).long()
+        return F.cross_entropy(self.logits, targets)
+
+    def release(self):
```
Comment (FelixAbrahamsson) [inline template/{{cookiecutter.repository_name}}/{{cookiecutter.package_name}}/architecture/prediction.py]: Perhaps `.cpu()` is a more descriptive name?

File: template/{{cookiecutter.repository_name}}/{{cookiecutter.package_name}}/architecture/simple_model.py [hunk 2]
```diff
@@ -1,23 +1,49 @@
...
+            nn.Dropout2d(0.25),
+            partial(torch.flatten, start_dim=1),
+            nn.Linear(9216, 128),
+            F.relu,
+            nn.Dropout2d(0.5),
+            nn.Linear(128, 10),
+            partial(F.log_softmax, dim=1),
+        )
+
+    def preprocessed(self, images):
```
Comment (FelixAbrahamsson) [inline template/{{cookiecutter.repository_name}}/{{cookiecutter.package_name}}/architecture/simple_model.py]: I'm not sure about having this as a class method, it's quite general. Maybe just a function inside `package.architecture`?

**Review by FelixAbrahamsson — COMMENTED** [review]
I like these ideas, however I'm not sure we should merge them into ml-workflow before testing them out on a real scenario. Initial concerns:
- Moving collate into the model architecture is another deviation from the norm. I like the flexibility, but it raises the threshold a bit for a new user as well as forces us to change the default collate fn when using a torch data loader.
- The prediction objects look neat on the surface but I'm not sure it will be convenient for every use-case, for example if the batches are complex and contain a lot of different input tensors/target tensors. Maybe it will be amazing, maybe it wont, we should test it out a bit.

File: template/{{cookiecutter.repository_name}}/{{cookiecutter.package_name}}/architecture/prediction.py (hunk 1 above)
Comment (samedii) [reply template/{{cookiecutter.repository_name}}/{{cookiecutter.package_name}}/architecture/prediction.py]: We need to detach too though?

File: template/{{cookiecutter.repository_name}}/{{cookiecutter.package_name}}/architecture/simple_model.py (hunk 2 above)
Comment (samedii) [reply template/{{cookiecutter.repository_name}}/{{cookiecutter.package_name}}/architecture/simple_model.py]: It is only meant to be used by `model` so I think it should be inside model.py

The reason I put it as a method is to allow adversarial examples easily in `train_batch` as we discussed yesterday.

Something like this
```
def train_batch(...):
    preprocessed = torch.tensor(
        model.preprocessed(images), requires_grad=True
    )
    with requires_nograd(model):
        prediction = model.core(preprocessed)
    prediction.loss().backward()
    prediction = model.core(adversarial(preprocessed.detach()))
```

File: template/{{cookiecutter.repository_name}}/{{cookiecutter.package_name}}/architecture/prediction.py (hunk 1 above)
Comment (FelixAbrahamsson) [reply template/{{cookiecutter.repository_name}}/{{cookiecutter.package_name}}/architecture/prediction.py]: Yeah but that seems fine to hide to me. Or do you disagree?

File: template/{{cookiecutter.repository_name}}/{{cookiecutter.package_name}}/architecture/simple_model.py (hunk 2 above)
Comment (FelixAbrahamsson) [reply template/{{cookiecutter.repository_name}}/{{cookiecutter.package_name}}/architecture/simple_model.py]: Ah, you're right about that. Perhaps if we note that we reuse the same code in multiple models it could be in a separate function that the models import and use in their own `self.preprocess`.

File: template/{{cookiecutter.repository_name}}/{{cookiecutter.package_name}}/architecture/prediction.py (hunk 1 above)
Comment (samedii) [reply template/{{cookiecutter.repository_name}}/{{cookiecutter.package_name}}/architecture/prediction.py]: Yes I dislike it. I can create a separate cpu and detach instead

**Comment by Jim-Holmstroem** [comment]
we could move them to a module called experimental and either remove it or move them out of there when we feel that feature is ready for the mainstream
(we just have to make sure that everything is marked as experimental in documentation and that the code gives one warning of "using experimental features" when started, this can be done with an `@experimental` decorator

**Comment by samedii** [comment]
I don't see how we would isolate this to "experimental". I've tried it in another small project and it worked well there. We can maybe try it on ampli as well

**Comment by samedii** [comment]
I think we should merge as this is pretty well tried now. If we find a better way then we can just change this later. This does not cause any back compatibility issues since it's just the template

**Review by Jim-Holmstroem — APPROVED** [review]

**Comment by Jim-Holmstroem** [comment]
(when the tests pass)

**Review by Jim-Holmstroem — APPROVED** [review]

**Comment by FelixAbrahamsson** [comment]
After trying out this workflow on Ampli I quite like it :)
The only thing I'm considering is the `model.forward`/`model.preprocessed` split. Maybe it doesn't belong in the default template as it's (as far as we know) only in special cases that we need to be able to store the preprocessed input or be able to backpropagate all the way to the input (GANs, adversarial input).
What do you think about this?

**Comment by samedii** [comment]
I think it makes sense that "forward" can be backprop:ed which motivates the split for me. I also think that the logic can get very long and this is a natural way of splitting the code into multiple parts.
